README.TXT
This is the fork of MikeOS, customized. Has bootlogo. Graphical bootlogo. The OS for COMPuters, not for "Cool hot things with OVER 9000 GHZ and DDR3" :) 
Original read me below.
Sincerely,
~ Vladislav Korotnev
==================================================================
MikeOS -- Open source 16-bit operating system for x86 PCs
Copyright (C) 2006 - 2011 MikeOS Developers -- see doc/LICENSE.TXT
==================================================================


MikeOS is a 16-bit real mode operating system for x86-compatible PCs, 
written entirely in assembly language, which boots from a floppy disk,
CD-ROM or USB key. It features a text-based dialog-driven user
interface, a command-line, support for FAT12 (MS-DOS-like) floppy
disks, sound (PC speaker), text editor, BASIC interpreter and more.
The kernel includes over 60 system calls.

MikeOS is a learning tool for those wishing to understand simple OS 
construction and x86 assembly. Quick getting-started guide: MikeOS can 
run from a floppy disk or CD-ROM, either on an emulator or a real PC. 
See the disk_images/ directory for files that you can write to the 
approprate media or boot in, for instance, VMware, QEMU or VirtualBox.

You can find the source code in the source/ directory, and sample 
programs (included on the disk images) in the programs/ directory.
See the doc/ directory for more info, including:

 # handbook-user.html -- How to run and use MikeOS

 # handbook-appdev-basic.html -- Writing software in BASIC
 # handbook-appdev-asm.html -- Writing software in assembly
 # handbook-sysdev.html -- Building and modifying the OS

 # LICENSE.TXT -- The open source, BSD-like license
 # CHANGES.TXT -- Detailed list of changes in previous releases
 # CREDITS.TXT -- People involved in the project

Have fun, and see the website at http://mikeos.berlios.de

 -- Mike Saunders (okachi@gmail.com)


==================================================================



doc/CHANGES.TXT
==================================================================
MikeOS -- Open source 16-bit operating system for x86 PCs
Copyright (C) 2006 - 2011 MikeOS Developers -- see doc/LICENSE.TXT
==================================================================


4.1.1 (20 January 2011)
-----------------------

# source/features/screen.asm: os_list_dialog now scrollable for more
  entries than can fit into the dialog box

# source/features/cli.asm: dirlist expanded from 255 to 1024 bytes

# source/features/screen.asm: os_file_selector: buffer expanded to
  1024 bytes

# Build scripts now copy all .bas files in programs/ to the floppy
  disk image (and not just example.bas)

# source/features/misc.asm: os_pause rewritten by Paulo Valongo to work
  with more BIOSes; now takes 110ms chunks in AX

# source/features/disk.asm fixes from Mike Gonta: os_get_file_list now
  checks for directory entries AND volume label; os_write_file routine
  now properly zeroes out the .free_clusters list

# source/features/keyboard.asm: os_check_for_key now returns 0 in AX
  (and not just AL) when there's no key, as per the API docs

# Documentation and comment fixes



4.1 (17 January 2010)
---------------------

# BASIC variable assignment enhanced: can now take multiple values
  and variables, eg "x = a + 2 * b + 3" (thanks Steve)

# Two new system calls: os_port_byte_out and os_port_byte_in --
  send/receive bytes from ports (thanks Paulo)

# os_serial_port_enable system call: now takes 0 in AX for normal
  mode (9600 baud as before), or 1 for a new slow mode (1200 baud)

# New SERIAL command in BASIC: "SERIAL ON 1200" or "SERIAL ON 9600"
  to enable, "SERIAL SEND X" or "SERIAL SEND 50" to send a byte,
  and "SERIAL REC X" to receive a byte into a variable

# New PORT command in BASIC: "PORT OUT 1234 X" or "PORT OUT 1234 20"
  to send byte to a port, and "PORT IN 1234 X" to receive a byte

# source/features/serial.asm renamed to source/features/ports.asm
  as it now contains new non-serial routines

# os_get_random rewritten: now takes low and high values in AX and
  BX, and returns random int between them (inclusive) in CX; also
  has new os_seed_random routine used by the kernel (thanks Ian)

# New RAND command in BASIC: generates a random number between two
  values and stores the result in a variable

# CLI command checking code simplified -- no need to point SI to
  the input string each time (thanks Ian)

# BASIC CURSORCHAR command is now CURSCHAR, so that it doesn't
  look like a superset of the CURSOR command

# API version bumped to 13

# Handbook updates and fixes



4.0 (20 November 2009)
----------------------

# MikeOS BASIC included in the kernel; run EXAMPLE.BAS to try
  it out, and see doc/handbook-appdev-basic.html for a guide

# os_run_basic call accessible to use programs: takes AX as start
  of code location, and BX as size of the code (in bytes)

# App menu and CLI can now run .BAS programs as well as .BIN

# EDIT.BIN can edit .BAS files; press F8 to run the
  currently-being-edited code in the BASIC intepreter

# EDIT.BIN: F5 deletes whole lines; Backspace key now functions,
  although newlines still need to be removed with Delete; Esc
  quits without saving -- press F2 to save (as per bottom bar)

# If AUTORUN.BIN or AUTORUN.BAS exists on the disk, it will be
  executed immediately after the kernel has loaded, before showing
  the normal program launcher menu

# EDIT.BIN now pops up an error dialog if it can't save a file

# source/features/string.asm: os_string_strip rewritten -- shorter
  and saves 256 bytes (thanks to Ian Seyler)

# source/features/misc.asm: rewrote os_pause routine -- it now takes
  AX = tenths of a second to wait (so eg AX = 30 = 3 seconds)

# programs/hello.asm and programs/gfxdemo.asm removed (and their binaries
  on the floppy) -- old and not useful now

# os_dialog_box no longer raises a fatal error if strings are too long to
  fit in the box

# source/features/disk.asm: os_write_file no longer overwrites a file
  if it already exists; instead it bails out (carry flag set)

# API version bumped to 12

# source/features/disk.asm: filename checking and file writing bugfixes

# doc/handbook-appdev.html renamed doc/handbook-appdev-asm.html



3.3 (20 February 2009)
----------------------

# New program: EDIT.BIN -- a full-screen text editor (for Unix-formatted
  text files)

# README.TXT is no longer copied to the disk images; instead,
  programs/sample.txt is added for use in EDIT.BIN

# CLI now includes CAT command to display a file: 'CAT FOO.TXT'

# Music keyboard (KEYBOARD.BIN) given new interface to show a
  visual representation of a music keyboard

# os_write_file now removes an older version of the specified file,
  if it exists, before performing the write operation

# CLI code (source/features/cli.asm) cleaned up -- split into sections
  with fewer local labels

# os_string_parse call is now accessible to external programs

# Disk calls now convert a filename to uppercase automatically, so
  there's no need for the calling code to do it

# API version bumped to 11



3.2 (25 January 2009)
---------------------

# New system call: os_sint_to_string from Tslil -- converts a
  signed integer to a string

# os_int_to_string rewritten by Tslil; cleaner code, and no longer
  needs to be passed BX for a string location (it returns a pointer
  to an internal string via AX)

# New system call: os_serial_port_enable -- initialises the port

# Serial port is no longer initialised by the kernel, but by any
  external programs that need it (eg in programs/serial.asm)

# New system call: os_string_reverse (source/features/string.asm)

# os_filename_convert removed; source/features/disk.asm now does
  the conversion internally

# API version bumped to 10

# Handbooks updated with API changes


source/kernel.asm
; ==================================================================
; MikeOS -- The Mike Operating System kernel
; Copyright (C) 2006 - 2011 MikeOS Developers -- see doc/LICENSE.TXT
;
; This is loaded from the floppy/CD, by BOOTLOAD.BIN, as KERNEL.BIN.
; First we have the system call vectors, which start at a static point
; for programs to jump to. Following that is the main kernel code and
; then additional system call code is included.
; ==================================================================


	BITS 16

	%DEFINE MIKEOS_VER '1.1'	; OS version number
	%DEFINE MIKEOS_API_VER 13	; API version for programs to check


	; This is the location in RAM for kernel disk operations, 24K
	; after the point where the kernel has loaded; it's 8K in size,
	; because external programs load after it at the 32K point:

	disk_buffer	equ	24576


; ------------------------------------------------------------------
; OS CALL VECTORS -- Static locations for system call vectors
; Note: these cannot be moved, or it'll break the calls!

; The comments show exact locations of instructions in this section,
; and are used in programs/mikedev.inc so that an external program can
; use a MikeOS system call without having to know its exact position
; in the kernel source code...

os_call_vectors:
	jmp os_main			; 0000h -- Called from bootloader
	jmp os_print_string		; 0003h
	jmp os_move_cursor		; 0006h
	jmp os_clear_screen		; 0009h
	jmp os_print_horiz_line		; 000Ch
	jmp os_print_newline		; 000Fh
	jmp os_wait_for_key		; 0012h
	jmp os_check_for_key		; 0015h
	jmp os_int_to_string		; 0018h
	jmp os_speaker_tone		; 001Bh
	jmp os_speaker_off		; 001Eh
	jmp os_load_file		; 0021h
	jmp os_pause			; 0024h
	jmp os_fatal_error		; 0027h
	jmp os_draw_background		; 002Ah
	jmp os_string_length		; 002Dh
	jmp os_string_uppercase		; 0030h
	jmp os_string_lowercase		; 0033h
	jmp os_input_string		; 0036h
	jmp os_string_copy		; 0039h
	jmp os_dialog_box		; 003Ch
	jmp os_string_join		; 003Fh
	jmp os_get_file_list		; 0042h
	jmp os_string_compare		; 0045h
	jmp os_string_chomp		; 0048h
	jmp os_string_strip		; 004Bh
	jmp os_string_truncate		; 004Eh
	jmp os_bcd_to_int		; 0051h
	jmp os_get_time_string		; 0055h
	jmp os_get_api_version		; 0057h
	jmp os_file_selector		; 005Ah
	jmp os_get_date_string		; 005Dh
	jmp os_send_via_serial		; 0060h
	jmp os_get_via_serial		; 0063h
	jmp os_find_char_in_string	; 0066h
	jmp os_get_cursor_pos		; 0069h
	jmp os_print_space		; 006Ch
	jmp os_dump_string		; 006Fh
	jmp os_print_digit		; 0072h
	jmp os_print_1hex		; 0075h
	jmp os_print_2hex		; 0078h
	jmp os_print_4hex		; 007Bh
	jmp os_long_int_to_string	; 007Eh
	jmp os_long_int_negate		; 0081h
	jmp os_set_time_fmt		; 0084h
	jmp os_set_date_fmt		; 0087h
	jmp os_show_cursor		; 008Ah
	jmp os_hide_cursor		; 008Dh
	jmp os_dump_registers		; 0090h
	jmp os_string_strincmp		; 0093h
	jmp os_write_file		; 0096h
	jmp os_file_exists		; 0099h
	jmp os_create_file		; 009Ch
	jmp os_remove_file		; 009Fh
	jmp os_rename_file		; 00A2h
	jmp os_get_file_size		; 00A5h
	jmp os_input_dialog		; 00A8h
	jmp os_list_dialog		; 00ABh
	jmp os_string_reverse		; 00AEh
	jmp os_string_to_int		; 00B1h
	jmp os_draw_block		; 00B4h
	jmp os_get_random		; 00B7h
	jmp os_string_charchange	; 00BAh
	jmp os_serial_port_enable	; 00BDh
	jmp os_sint_to_string		; 00C0h
	jmp os_string_parse		; 00C3h
	jmp os_run_basic		; 00C6h
	jmp os_port_byte_out		; 00C9h
	jmp os_port_byte_in		; 00CCh


; ------------------------------------------------------------------
; START OF MAIN KERNEL CODE

os_main:
	cli				; Clear interrupts
	mov ax, 0
	mov ss, ax			; Set stack segment and pointer
	mov sp, 0FFFFh
	sti				; Restore interrupts

	cld				; The default direction for string operations
					; will be 'up' - incrementing address in RAM

	mov ax, 2000h			; Set all segments to match where kernel is loaded
	mov ds, ax			; After this, we don't need to bother with
	mov es, ax			; segments ever again, as MikeOS and its programs
	mov fs, ax			; live entirely in 64K
	mov gs, ax

	mov ax, 1003h			; Set text output with certain attributes
	mov bx, 0			; to be bright, and not blinking
	int 10h

	call os_seed_random		; Seed random number generator

	;BootLogo
	mov ax, logo_bin_file_name
	call os_file_exists
	jc no_autorun_bin		; Skip next three lines if LOGO.BIN doesn't exist

	mov cx, 32768			; Otherwise load the program into RAM...
	call os_load_file
	jmp execute_logo_program		; ...and move on to the executing part


autorun_check:
	; Let's see if there's a file called AUTORUN.BIN and execute
	; it if so, before going to the program launcher menu

	mov ax, autorun_bin_file_name
	call os_file_exists
	jc no_autorun_bin		; Skip next three lines if AUTORUN.BIN doesn't exist

	mov cx, 32768			; Otherwise load the program into RAM...
	call os_load_file
	jmp execute_bin_program		; ...and move on to the executing part


	; Or perhaps there's an AUTORUN.BAS file?

no_autorun_bin:
	mov ax, autorun_bas_file_name
	call os_file_exists
	jc option_screen		; Skip next section if AUTORUN.BAS doesn't exist

	mov cx, 32768			; Otherwise load the program into RAM
	call os_load_file
	call os_clear_screen
	mov ax, 32768
	call os_run_basic		; Run the kernel's BASIC interpreter

	jmp app_selector		; And go to the app selector menu when BASIC ends


	; Now we display a dialog box offering the user a choice of
	; a menu-driven program selector, or a command-line interface

option_screen:
	mov ax, os_init_msg		; Set up the welcome screen
	mov bx, os_version_msg
	mov cx, 00001111b		; Colour: white text on light blue
	call os_draw_background

	mov ax, dialog_string_1		; Ask if user wants app selector or command-line
	mov bx, dialog_string_2
	mov cx, dialog_string_3
	mov dx, 1			; We want a two-option dialog box (OK or Cancel)
	call os_dialog_box

	cmp ax, 1			; If OK (option 0) chosen, start app selector
	jne near app_selector

	call os_clear_screen		; Otherwise clean screen and start the CLI
	call os_command_line

	jmp option_screen		; Offer menu/CLI choice after CLI has exited


	; Data for the above code...

	os_init_msg		db 'Welcome to CompOS', 0
	os_version_msg		db 'Version ', MIKEOS_VER, 0

	dialog_string_1		db 'Thanks for trying out CompOS!', 0
	dialog_string_2		db 'Please select an interface: OK for the', 0
	dialog_string_3		db 'program menu, Cancel for command line.', 0



app_selector:
	mov ax, os_init_msg		; Draw main screen layout
	mov bx, os_version_msg
	mov cx, 00001111b		; Colour: white text on light blue
	call os_draw_background

	call os_file_selector		; Get user to select a file, and store
					; the resulting string location in AX
					; (other registers are undetermined)

	jc option_screen		; Return to the CLI/menu choice screen if Esc pressed

	mov si, ax			; Did the user try to run 'KERNEL.BIN'?
	mov di, kern_file_name
	call os_string_compare
	jc no_kernel_execute		; Show an error message if so

	mov si, ax			; Did the user try to run 'LOGO.BIN'?
	mov di, logo_bin_file_name
	call os_string_compare
	jc no_logo_execute		; Show an error message if so


	; Next, we need to check that the program we're attempting to run is
	; valid -- in other words, that it has a .BIN extension

	push si				; Save filename temporarily

	mov bx, si
	mov ax, si
	call os_string_length

	mov si, bx
	add si, ax			; SI now points to end of filename...

	dec si
	dec si
	dec si				; ...and now to start of extension!

	mov di, bin_ext
	mov cx, 3
	rep cmpsb			; Are final 3 chars 'BIN'?
	jne not_bin_extension		; If not, it might be a '.BAS'

	pop si				; Restore filename


	mov ax, si
	mov cx, 32768			; Where to load the program file
	call os_load_file		; Load filename pointed to by AX


execute_bin_program:
	call os_clear_screen		; Clear screen before running

	mov ax, 0			; Clear all registers
	mov bx, 0
	mov cx, 0
	mov dx, 0
	mov si, 0
	mov di, 0

	call 32768			; Call the external program code,
					; loaded at second 32K of segment
					; (program must end with 'ret')

	call os_clear_screen		; When finished, clear screen
	jmp app_selector		; and go back to the program list

execute_logo_program:
	call os_clear_screen		; Clear screen before running

	mov ax, 0			; Clear all registers
	mov bx, 0
	mov cx, 0
	mov dx, 0
	mov si, 0
	mov di, 0

	call 32768			; Call the external program code,
					; loaded at second 32K of segment
					; (program must end with 'ret')

	call os_clear_screen		; When finished, clear screen
	jmp autorun_check



no_kernel_execute:			; Warn about trying to executing kernel!
	mov ax, kerndlg_string_1
	mov bx, kerndlg_string_2
	mov cx, kerndlg_string_3
	mov dx, 0			; One button for dialog box
	call os_dialog_box

	jmp app_selector		; Start over again...


no_logo_execute:			; Warn about trying to executing kernel!
	mov ax, logodlg_string_1
	mov bx, logodlg_string_2
	mov cx, logodlg_string_3
	mov dx, 0			; One button for dialog box
	call os_dialog_box

	jmp app_selector		; Start over again...

not_bin_extension:
	pop si				; We pushed during the .BIN extension check

	push si				; Save it again in case of error...

	mov bx, si
	mov ax, si
	call os_string_length

	mov si, bx
	add si, ax			; SI now points to end of filename...

	dec si
	dec si
	dec si				; ...and now to start of extension!

	mov di, bas_ext
	mov cx, 3
	rep cmpsb			; Are final 3 chars 'BAS'?
	jne not_bas_extension		; If not, error out


	pop si

	mov ax, si
	mov cx, 32768			; Where to load the program file
	call os_load_file		; Load filename pointed to by AX

	call os_clear_screen		; Clear screen before running

	mov ax, 32768
	call os_run_basic		; And run our BASIC interpreter on the code!

	mov si, basic_finished_msg
	call os_print_string
	call os_wait_for_key

	call os_clear_screen
	jmp app_selector		; and go back to the program list


not_bas_extension:
	pop si

	mov ax, ext_string_1
	mov bx, ext_string_2
	mov cx, 0
	mov dx, 0			; One button for dialog box
	call os_dialog_box

	jmp app_selector		; Start over again...


	; And now data for the above code...

	kern_file_name		db 'KERNEL.BIN', 0

	autorun_bin_file_name	db 'AUTORUN.BIN', 0
	logo_bin_file_name	db 'LOGO.BIN', 0
	autorun_bas_file_name	db 'AUTORUN.BAS', 0

	bin_ext			db 'BIN'
	bas_ext			db 'BAS'

	kerndlg_string_1	db 'Cannot load and execute CompOS kernel!', 0
	kerndlg_string_2	db 'KERNEL.BIN is the core of CompOS, and', 0
	kerndlg_string_3	db 'is not a normal program.', 0

	logodlg_string_1	db 'Cannot load and execute CompOS logo!', 0
	logodlg_string_2	db 'LOGO.BIN is the logo of CompOS, and', 0
	logodlg_string_3	db 'is not a normal program.', 0

	ext_string_1		db 'Invalid filename extension! You can', 0
	ext_string_2		db 'only execute .BIN or .BAS programs.', 0

	basic_finished_msg	db '>>> CompOS BASIC program finished -- press a key', 0


; ------------------------------------------------------------------
; SYSTEM VARIABLES -- Settings for programs and system calls


	; Time and date formatting

	fmt_12_24	db 0		; Non-zero = 24-hr format

	fmt_date	db 0, '/'	; 0, 1, 2 = M/D/Y, D/M/Y or Y/M/D
					; Bit 7 = use name for months
					; If bit 7 = 0, second byte = separator character


; ------------------------------------------------------------------
; FEATURES -- Code to pull into the kernel


	%INCLUDE "features/cli.asm"
	%INCLUDE "features/basic.asm"
 	%INCLUDE "features/disk.asm"
	%INCLUDE "features/keyboard.asm"
	%INCLUDE "features/math.asm"
	%INCLUDE "features/misc.asm"
	%INCLUDE "features/ports.asm"
	%INCLUDE "features/screen.asm"
	%INCLUDE "features/sound.asm"
	%INCLUDE "features/string.asm"


; ==================================================================
; END OF KERNEL
; ==================================================================

